scripts/utils/CLI.ts

```diff
@@ -0,0 +1,299 @@
+ */
+type PositionalArg<T = unknown> = StringArg<T> & {
+    name: string;
+};
+
+/**
+ * This type represents the config for a CLI.
+ * Note: this utility does not yet support variadic args of any kind.
+ */
+type CLIConfig = NonEmptyObject<{
+    /**
+     * Record of named flags that are fully characterized by their presence or absence (present=true,absent=false).
+     * @example `--verbose`
+     */
+    flags?: Record<string, BooleanArg>;
+
+    /**
+     * Record of named arguments that are represented by a key and a value.
+     * @example `--threads=8`
+     * @example `--name Rory`
+     */
+    namedArgs?: Record<string, StringArg>;
+
+    /**
+     * Tuple of positional args.
+     * @example `myScript.ts arg1 arg2 arg3`
+     */
+    positionalArgs?: NonEmptyTuple<PositionalArg>;
+}>;
+
+/**
+ * Record of flags to boolean after parsing.
+ */
+type ParsedFlags<Flags extends CLIConfig['flags']> = {
+    [K in keyof NonNullable<Flags>]: boolean;
+};
+
+/**
+ * Utility type to infer the final value of a string param. Either:
+ * - it's a plain string, or
+ * - it has a parse function and the final value is inferred from the return type of that function
+ */
+type InferStringArgParsedValue<T extends StringArg> = T extends {parse: (val: string) => infer R} ? R : string;
+
+/**
+ * Record of named args after parsing.
+ */
+type ParsedNamedArgs<NamedArgs extends CLIConfig['namedArgs']> = {
+    [K in keyof NonNullable<NamedArgs>]: InferStringArgParsedValue<NonNullable<NamedArgs>[K]>;
+};
+
+/**
+ * Record of positional args after parsing.
+ */
+type ParsedPositionalArgs<PositionalArgs extends CLIConfig['positionalArgs']> = {
+    [K in NonNullable<PositionalArgs>[number] as K['name']]: InferStringArgParsedValue<K>;
+};
+
+/**
+ * Utility to parse command-line arguments to a script.
+ *
+ * @example
+ * ```
+ * const cli = new CLI({
+ *     flags: {
+ *         verbose: {
+ *             description: 'Enable verbose logging',
+ *         },
+ *     },
+ *     namedArgs: {
+ *         time: {
+ *             description: 'Time of day to greet (morning or evening)',
+ *             default: 'morning',
+ *             parse: (val) => {
+ *                 if (val !== 'morning' && val !== 'evening') {
+ *                     throw new Error('Must be "morning" or "evening"');
+ *                 }
+ *                 return val as 'morning' | 'evening';
+ *             },
+ *         },
+ *     },
+ *     positionalArgs: [
+ *         {
+ *             name: 'firstName'
+ *             description: 'First name to greet',
+ *         },
+ *         {
+ *             name: 'lastName',
+ *             description: 'Last name to greet',
+ *             default: '',
+ *         },
+ *     ],
+ * });
+ *
+ * let fullName = cli.parameters.firstName;
+ * if (cli.flags.verbose) {
+ *     fullName += cli.parameters.lastName;
+ * }
+ * console.log(fullName);
+ * console.log(cli.flags.time);
```

Comment:
Super minor thing 😄 I think that according to the example it should be:
```suggestion
 * console.log(cli.namedArgs.time);
```

Reply:
good catch